File: core/02 - Tutorials/07 - Transaction Sign/TransactionSign_ViewController.m
/*
 * Copyright © 2020-2023 THALES. All rights reserved.
 */

// IMPORTANT: This source code is intended to serve training information purposes only.
//            Please make sure to review our IdCloud documentation, including security guidelines.

#import "TransactionSign_ViewController.h"
#import "Provisioning_Logic.h"
#import "TransactionSign_Logic.h"

@interface TransactionSign_ViewController() <UITextFieldDelegate>

@property (weak, nonatomic) IBOutlet UITextField *textAmount;
@property (weak, nonatomic) IBOutlet UITextField *textBeneficiary;

@end

// Review: SNE document private methods
@implementation TransactionSign_ViewController

// MARK: - Lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Used to handle return key.
    [_textAmount        setDelegate:self];
    [_textBeneficiary   setDelegate:self];
}

// MARK: - Shared

- (void)generateAndDisplaySignatureOtp:(id<EMOathToken>)token
                             authInput:(id<EMAuthInput>)authInput
                                amount:(NSString *)amount
                           beneficiary:(NSString *)beneficiary
{
    // Calculate OTP for given pin and display result.
    NSError     *internalErr    = nil;
    OTP_Value   *otpValue       = [TransactionSign_Logic generateOtp:token
                                                       withAuthInput:authInput
                                                              amount:amount
                                                         beneficiary:beneficiary
                                                               error:&internalErr];
    
    if (internalErr) {
        [self displayMessageDialogError:internalErr];
    } else {
        [self displayMessageResult:otpValue.otp.stringValue lifespan:otpValue.lifespan];
    }

    // All secure containers should be wiped asap.
    [authInput wipe];
    [otpValue wipe];
}

// MARK: - UITextFieldDelegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    // Hide current keyboard.
    [textField resignFirstResponder];
    
    // Jump to next one.
    if ([textField isEqual:_textAmount]) {
        [_textBeneficiary becomeFirstResponder];
    }
    
    // All actions are allowed
    return YES;
}

// MARK: - User interface

- (IBAction)onButtonPressedGenerateSignatureOtp:(UIButton *)sender
{
    [self userPinWithCompletionHandler:^(id<EMPinAuthInput> pin) {
        [self generateAndDisplaySignatureOtp:[Provisioning_Logic token]
                                   authInput:pin
                                      amount:self.textAmount.text
                                 beneficiary:self.textBeneficiary.text];
    }];
}


@end

File: fasttrack/02 - Tutorials/07 - Transaction Sign/TransactionSignViewController.swift
/*
 * Copyright © 2020-2023 THALES. All rights reserved.
 */

// IMPORTANT: This source code is intended to serve training information purposes only.
//            Please make sure to review our IdCloud documentation, including security guidelines.


class TransactionSignViewController: ChangePinViewController, UITextFieldDelegate {
    @IBOutlet weak var textAmountTextField: UITextField!
    @IBOutlet weak var textBeneficiaryTextField: UITextField!
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        textAmountTextField.delegate = self
        textBeneficiaryTextField.delegate = self
    }
    
    func generateOtp(_ token: EMProtectorOathTokenDevice,
                     pinAuthInput: EMProtectorAuthInput?,
                     pin: String?,
                     amount: String, beneficiary: String) {
        var otpValue: OtpValue?
        do {
            otpValue = try TransactionSignLogic.generateOtp(token, pinAuthInput: pinAuthInput, pin: pin, amount: amount, beneficiary: beneficiary)
            displayOTP(value: otpValue!.otp, lifespan: otpValue!.lifespan)
        } catch let error as NSError {
            displayMessageDialogError(error: error)
        }
    }
    
    @IBAction func buttonGenerateSignatureOTPPressed(_ sender: Any) {
        let oathProtector = ProvisioningLogic.getOathProtector()
        let ocraSettings = EMProtectorOcraSettings.init()
        ocraSettings.ocraSuite = OtpConfig.getOcraSuite()
        var token:EMProtectorOathTokenDevice?
        do {
            if let tokenNames: Set<String> = try? oathProtector?.tokenDeviceNames(), !tokenNames.isEmpty {
                token = try? oathProtector?.tokenDevice(name: tokenNames.first!, fingerprintCustomData: ProvisioningConfig.getCustomFingerprintData(), ocraSettings: ocraSettings)
            }
        }
        
        showPinInput { [self] authInput, pin in
            generateOtp(token!, pinAuthInput: authInput, pin: pin, amount: textAmountTextField.text!, beneficiary: textBeneficiaryTextField.text!)
        }
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        if textField.isEqual(self.textAmountTextField) {
            self.textBeneficiaryTextField .becomeFirstResponder()
        }
        return true
    }
}
